### ISSUE TITLE
ICE when using Bevy's Taskpool

### ISSUE THREAD

**Issue by Kjolnyr** [rust-lang/rust#110258]
My Cargo.toml
```toml
[package]
name = "rust_bug"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
bevy = { version = "0.10.1" }
```

### Code

```Rust
use bevy::tasks::{Scope, TaskPool};

fn main() {
    let pool = TaskPool::new();
    pool.scope(|_s: Scope<()>| {});
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.68.1 (8460ca823 2023-03-20)
binary: rustc
commit-hash: 8460ca823e8367a30dda430efda790588b8c84d3
commit-date: 2023-03-20
host: x86_64-pc-windows-msvc
release: 1.68.1
LLVM version: 15.0.6
```

### Error output

```
error: internal compiler error: compiler\rustc_infer\src\infer\region_constraints\mod.rs:568:17: cannot relate bound region: ReLateBound(DebruijnIndex(0), BoundRegion { var: 0, kind: BrNamed(DefId(0:8 ~ rust_bug[783e]::main::'_), '_) }) <= ReLateBound(DebruijnIndex(0), BoundRegion { var: 0, kind: BrNamed(DefId(42:101 ~ bevy_tasks[e101]::task_pool::{impl#2}::scope::'scope), 'scope) })

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/8460ca823e8367a30dda430efda790588b8c84d3\compiler\rustc_errors\src\lib.rs:987:33
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0:     0x7ff9b6239d02 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h24420960574ebc1c
   1:     0x7ff9b627570b - core::fmt::write::h26edb44d1c5a6228
   2:     0x7ff9b622cdfa - <std::io::IoSlice as core::fmt::Debug>::fmt::h3ff7218cf0d77c3d
   3:     0x7ff9b6239a4b - std::sys::common::alloc::realloc_fallback::hc95eb280a5e906e1
   4:     0x7ff9b623d3f9 - std::panicking::default_hook::h0e3996dde5e60121
   5:     0x7ff9b623d07b - std::panicking::default_hook::h0e3996dde5e60121
   6:     0x7ff98f3968d5 - rustc_driver[9a81db59c34cf3a5]::describe_lints
   7:     0x7ff9b623dd5f - std::panicking::rust_panic_with_hook::hf3035ce96f91938a
   8:     0x7ff991656493 - <<rustc_infer[4a5594bed6e23f7a]::infer::error_reporting::TypeErrCtxt>::construct_generic_bound_failure::SubOrigin as core[88a2d7387c40f712]::fmt::Debug>::fmt
   9:     0x7ff9916549a9 - <<rustc_infer[4a5594bed6e23f7a]::infer::error_reporting::TypeErrCtxt>::construct_generic_bound_failure::SubOrigin as core[88a2d7387c40f712]::fmt::Debug>::fmt
  10:     0x7ff991654619 - <<rustc_infer[4a5594bed6e23f7a]::infer::error_reporting::TypeErrCtxt>::construct_generic_bound_failure::SubOrigin as core[88a2d7387c40f712]::fmt::Debug>::fmt
  11:     0x7ff991654609 - <<rustc_infer[4a5594bed6e23f7a]::infer::error_reporting::TypeErrCtxt>::construct_generic_bound_failure::SubOrigin as core[88a2d7387c40f712]::fmt::Debug>::fmt
  12:     0x7ff991653fb0 - <<rustc_infer[4a5594bed6e23f7a]::infer::error_reporting::TypeErrCtxt>::construct_generic_bound_failure::SubOrigin as core[88a2d7387c40f712]::fmt::Debug>::fmt
  13:     0x7ff991653e09 - <<rustc_infer[4a5594bed6e23f7a]::infer::error_reporting::TypeErrCtxt>::construct_generic_bound_failure::SubOrigin as core[88a2d7387c40f712]::fmt::Debug>::fmt
  14:     0x7ff99165c259 - <rustc_infer[4a5594bed6e23f7a]::infer::opaque_types::table::OpaqueTypeStorage as core[88a2d7387c40f712]::fmt::Debug>::fmt
  15:     0x7ff99165c2bd - <rustc_infer[4a5594bed6e23f7a]::infer::opaque_types::table::OpaqueTypeStorage as core[88a2d7387c40f712]::fmt::Debug>::fmt
  16:     0x7ff99165b47a - <rustc_infer[4a5594bed6e23f7a]::infer::opaque_types::table::OpaqueTypeStorage as core[88a2d7387c40f712]::fmt::Debug>::fmt
  17:     0x7ff99165b408 - <rustc_infer[4a5594bed6e23f7a]::infer::opaque_types::table::OpaqueTypeStorage as core[88a2d7387c40f712]::fmt::Debug>::fmt
  18:     0x7ff99165b3c6 - <rustc_infer[4a5594bed6e23f7a]::infer::opaque_types::table::OpaqueTypeStorage as core[88a2d7387c40f712]::fmt::Debug>::fmt
  19:     0x7ff98f1c4d1a - <rustc_infer[4a5594bed6e23f7a]::infer::combine::ConstInferUnifier as rustc_middle[348c050efaf3cf62]::ty::relate::TypeRelation>::consts
  20:     0x7ff98f1cccbc - <rustc_middle[348c050efaf3cf62]::ty::sty::AliasTy as rustc_infer[4a5594bed6e23f7a]::infer::at::ToTrace>::to_trace
  21:     0x7ff98f1d3840 - <rustc_middle[348c050efaf3cf62]::ty::sty::AliasTy as rustc_infer[4a5594bed6e23f7a]::infer::at::ToTrace>::to_trace
  22:     0x7ff98f18ee99 - <rustc_infer[4a5594bed6e23f7a]::infer::equate::Equate as rustc_middle[348c050efaf3cf62]::ty::relate::TypeRelation>::tys
  23:     0x7ff98d95f4ea - rustc_trait_selection[ba7b18a923262ca4]::traits::const_evaluatable::is_const_evaluatable
  24:     0x7ff98d94c83b - <rustc_trait_selection[ba7b18a923262ca4]::traits::project::AssocTypeNormalizer as rustc_middle[348c050efaf3cf62]::ty::fold::TypeFolder>::fold_const
  25:     0x7ff9915a4e58 - <rustc_trait_selection[ba7b18a923262ca4]::traits::const_evaluatable::satisfied_from_param_env::Visitor as rustc_middle[348c050efaf3cf62]::ty::visit::TypeVisitor>::visit_const
  26:     0x7ff991543c09 - <rustc_trait_selection[ba7b18a923262ca4]::traits::error_reporting::method_chain::CollectAllMismatches as rustc_middle[348c050efaf3cf62]::ty::relate::TypeRelation>::consts
  27:     0x7ff9915bd363 - <rustc_infer[4a5594bed6e23f7a]::infer::error_reporting::TypeErrCtxt as rustc_trait_selection[ba7b18a923262ca4]::traits::error_reporting::suggestions::TypeErrCtxtExt>::report_closure_arg_mismatch
  28:     0x7ff9915d087b - <rustc_infer[4a5594bed6e23f7a]::infer::error_reporting::TypeErrCtxt as rustc_trait_selection[ba7b18a923262ca4]::traits::error_reporting::TypeErrCtxtExt>::report_selection_error
  29:     0x7ff9915dd4a5 - <rustc_infer[4a5594bed6e23f7a]::infer::error_reporting::TypeErrCtxt as rustc_trait_selection[ba7b18a923262ca4]::traits::error_reporting::InferCtxtPrivExt>::report_fulfillment_error
  30:     0x7ff9915cd209 - <rustc_infer[4a5594bed6e23f7a]::infer::error_reporting::TypeErrCtxt as rustc_trait_selection[ba7b18a923262ca4]::traits::error_reporting::TypeErrCtxtExt>::report_fulfillment_errors
  31:     0x7ff98e504ccb - <rustc_hir_typeck[de1d69685c3524a8]::fn_ctxt::FnCtxt>::demand_coerce
  32:     0x7ff98e4aefd8 - <rustc_hir_typeck[de1d69685c3524a8]::inherited::Inherited as core[88a2d7387c40f712]::ops::deref::Deref>::deref
  33:     0x7ff98d57f049 - <rustc_query_impl[7254303e48199d54]::on_disk_cache::CacheEncoder as rustc_type_ir[1f8388642b2bc09c]::codec::TyEncoder>::encode_alloc_id
  34:     0x7ff98d5e5215 - <&[(rustc_middle[348c050efaf3cf62]::ty::Predicate, rustc_span[d3e736c3e38f3ea5]::span_encoding::Span)] as rustc_serialize[fb792d7000edb6]::serialize::Decodable<rustc_query_impl[7254303e48199d54]::on_disk_cache::CacheDecoder>>::decode
  35:     0x7ff98ed6fbdd - <&[(rustc_middle[348c050efaf3cf62]::ty::Clause, rustc_span[d3e736c3e38f3ea5]::span_encoding::Span)] as rustc_serialize[fb792d7000edb6]::serialize::Decodable<rustc_query_impl[7254303e48199d54]::on_disk_cache::CacheDecoder>>::decode
  36:     0x7ff98ec46da3 - <rustc_query_impl[7254303e48199d54]::Queries as rustc_middle[348c050efaf3cf62]::ty::query::QueryEngine>::as_any
  37:     0x7ff98e5580a3 - <rustc_hir_typeck[de1d69685c3524a8]::fn_ctxt::FnCtxt>::resolve_fully_qualified_call
  38:     0x7ff98d22bcb1 - <rustc_hir_typeck[de1d69685c3524a8]::inherited::Inherited>::build
  39:     0x7ff98d580e1f - <rustc_query_impl[7254303e48199d54]::on_disk_cache::CacheEncoder as rustc_type_ir[1f8388642b2bc09c]::codec::TyEncoder>::encode_alloc_id
  40:     0x7ff98d6041bf - <&[(rustc_middle[348c050efaf3cf62]::ty::Predicate, rustc_span[d3e736c3e38f3ea5]::span_encoding::Span)] as rustc_serialize[fb792d7000edb6]::serialize::Decodable<rustc_query_impl[7254303e48199d54]::on_disk_cache::CacheDecoder>>::decode
  41:     0x7ff98d6730d1 - <&[(rustc_middle[348c050efaf3cf62]::ty::Predicate, rustc_span[d3e736c3e38f3ea5]::span_encoding::Span)] as rustc_serialize[fb792d7000edb6]::serialize::Decodable<rustc_query_impl[7254303e48199d54]::on_disk_cache::CacheDecoder>>::decode
  42:     0x7ff98d552d72 - <rustc_query_impl[7254303e48199d54]::Queries as rustc_middle[348c050efaf3cf62]::ty::query::QueryEngine>::try_mark_green
  43:     0x7ff98d28a45a - <<dyn rustc_hir_analysis[38f54a96453c487e]::astconv::AstConv>::create_substs_for_ast_path::{closure#0}::SubstsForAstPathCtxt as rustc_hir_analysis[38f54a96453c487e]::astconv::CreateSubstsForGenericArgsCtxt>::inferred_kind
  44:     0x7ff98d282149 - rustc_hir_analysis[38f54a96453c487e]::check_crate
  45:     0x7ff98cb62275 - rustc_interface[80bde221feee50b9]::passes::analysis
  46:     0x7ff98d580a4f - <rustc_query_impl[7254303e48199d54]::on_disk_cache::CacheEncoder as rustc_type_ir[1f8388642b2bc09c]::codec::TyEncoder>::encode_alloc_id
  47:     0x7ff98d5ffe69 - <&[(rustc_middle[348c050efaf3cf62]::ty::Predicate, rustc_span[d3e736c3e38f3ea5]::span_encoding::Span)] as rustc_serialize[fb792d7000edb6]::serialize::Decodable<rustc_query_impl[7254303e48199d54]::on_disk_cache::CacheDecoder>>::decode
  48:     0x7ff98d6ec201 - <&[(rustc_middle[348c050efaf3cf62]::ty::Predicate, rustc_span[d3e736c3e38f3ea5]::span_encoding::Span)] as rustc_serialize[fb792d7000edb6]::serialize::Decodable<rustc_query_impl[7254303e48199d54]::on_disk_cache::CacheDecoder>>::decode
  49:     0x7ff98d55056f - <rustc_query_impl[7254303e48199d54]::Queries as rustc_middle[348c050efaf3cf62]::ty::query::QueryEngine>::try_mark_green
  50:     0x7ff98cb0d1a9 - <rustc_middle[348c050efaf3cf62]::ty::SymbolName as core[88a2d7387c40f712]::fmt::Debug>::fmt
  51:     0x7ff98cb1e7ba - rustc_driver[9a81db59c34cf3a5]::args::arg_expand_all
  52:     0x7ff98cb02ac1 - <unknown>
  53:     0x7ff98cb0cacb - <rustc_middle[348c050efaf3cf62]::ty::SymbolName as core[88a2d7387c40f712]::fmt::Debug>::fmt
  54:     0x7ff98cb0c319 - <rustc_middle[348c050efaf3cf62]::ty::SymbolName as core[88a2d7387c40f712]::fmt::Debug>::fmt
  55:     0x7ff98cb0404d - <rustc_middle[348c050efaf3cf62]::ty::SymbolName as core[88a2d7387c40f712]::fmt::Debug>::fmt
  56:     0x7ff9b625021c - std::sys::windows::thread::Thread::new::hd26bc42258b98de9
  57:     0x7ffa375a26bd - BaseThreadInitThunk
  58:     0x7ffa3814a9f8 - RtlUserThreadStart
```

</p>
</details>

**Comment by ChrisDenton** [comment]
Does this happen if you build with nightly?

**Comment by Kjolnyr** [comment]
It doesn't happen with nightly actually

```code
Compiling rust_bug v0.1.0 (C:\Users\void\Documents\rust_bug)
error[E0631]: type mismatch in closure arguments
   --> src\main.rs:5:10
    |
5   |     pool.scope(|_s: Scope<()>| {});
    |          ^^^^^ --------------- found signature defined here
    |          |
    |          expected due to this
    |
    = note: expected closure signature `for<'scope> fn(&'scope bevy::tasks::Scope<'scope, '_, _>) -> _`
note: required by a bound in `TaskPool::scope`
   --> C:\Users\void\.cargo\registry\src\index.crates.io-6f17d22bba15001f\bevy_tasks-0.10.1\src\task_pool.rs:280:12
    |
280 |         F: for<'scope> FnOnce(&'scope Scope<'scope, 'env, T>),
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `TaskPool::scope`
help: consider borrowing the argument
    |
5   |     pool.scope(|_s: &Scope<()>| {});
    |                     +

For more information about this error, try `rustc --explain E0631`.
error: could not compile `rust_bug` (bin "rust_bug") due to previous error
```

```
rustc 1.70.0-nightly (db0cbc48d 2023-03-26)
binary: rustc
commit-hash: db0cbc48d4aaa300713a95d9b317a365a474490c
commit-date: 2023-03-26
host: x86_64-pc-windows-msvc
release: 1.70.0-nightly
LLVM version: 16.0.0
```

**Comment by ChrisDenton** [comment]
Thanks for investigating! It looks like this will be fixed in rust 1.69, which is releasing next week.

I'll close this as completed but please feel free to report any other ICEs you see.
